packages/gatsby-source-prismic/src/builders/buildImageBaseFieldConfigMap.ts

```diff
@@ -1,76 +1,62 @@
+import * as gqlc from 'graphql-compose'
 import { Dependencies, PrismicAPIImageField } from '../types'
 
 const resolveUrl = (source: PrismicAPIImageField): string | null => source.url
 
-const resolveWidth = (source: PrismicAPIImageField): number | undefined =>
+/* const resolveWidth = (source: PrismicAPIImageField): number | undefined =>
   source.dimensions?.width
 
 const resolveHeight = (source: PrismicAPIImageField): number | undefined =>
-  source.dimensions?.height
+  source.dimensions?.height */
 
 export const buildImageBaseFieldConfigMap: RTE.ReaderTaskEither<
   Dependencies,
   never,
   gqlc.ObjectTypeComposerFieldConfigMapDefinition<PrismicAPIImageField, unknown>
 > = pipe(
-  RTE.asks((deps) => ({
-    alt: 'String',
-    copyright: 'String',
-    dimensions: deps.globalNodeHelpers.createTypeName('ImageDimensionsType'),
-    url: gatsbyImgix.createImgixUrlFieldConfig({
-      paramsInputType: deps.globalNodeHelpers.createTypeName(
-        'ImgixUrlParamsInput',
-      ),
-      resolveUrl,
-      defaultImgixParams: deps.pluginOptions.imageImgixParams,
-    }),
-    fixed: gatsbyImgix.createImgixFixedFieldConfig({
-      type: deps.nodeHelpers.createTypeName('ImageFixedType'),
-      paramsInputType: deps.globalNodeHelpers.createTypeName(
-        'ImgixUrlParamsInput',
-      ),
-      resolveUrl,
-      resolveWidth,
-      resolveHeight,
+  RTE.asks((deps) => {
+    const imgixTypes = imgixGatsby.createImgixGatsbyTypes({
       cache: deps.cache,
-      defaultImgixParams: deps.pluginOptions.imageImgixParams,
-      defaultPlaceholderImgixParams:
-        deps.pluginOptions.imagePlaceholderImgixParams,
-    }),
-    fluid: gatsbyImgix.createImgixFluidFieldConfig({
-      type: deps.nodeHelpers.createTypeName('ImageFluidType'),
-      paramsInputType: deps.globalNodeHelpers.createTypeName(
-        'ImgixUrlParamsInput',
-      ),
       resolveUrl,
-      resolveWidth,
-      resolveHeight,
-      cache: deps.cache,
-      defaultImgixParams: deps.pluginOptions.imageImgixParams,
-      defaultPlaceholderImgixParams:
-        deps.pluginOptions.imagePlaceholderImgixParams,
-    }),
-    localFile: {
-      type: 'File',
-      resolve: async (
-        source: PrismicAPIImageField,
-      ): Promise<gatsbyFs.FileSystemNode | null> =>
-        source.url
-          ? await deps.createRemoteFileNode({
-              url: source.url,
-              store: deps.store,
-              cache: deps.cache,
-              createNode: deps.createNode,
-              createNodeId: deps.createNodeId,
-              reporter: deps.reporter,
-            })
-          : null,
-    },
-  })),
+      defaultParams: deps.pluginOptions.imageImgixParams,
+      namespace: 'Imgix',
+    })
+
+    const types = [
+      ...imgixTypes.types.map(deps.schema.buildObjectType),
+      ...imgixTypes.enumTypes.map(deps.schema.buildEnumType),
+      ...imgixTypes.inputTypes.map(deps.schema.buildInputObjectType),
+    ]
+    deps.createTypes(types)
```

Comment:
I've been using `ReaderTaskEither`'s `bind` to help organize everything into `pipe`s. A simple example looks like this: https://github.com/angeloashmore/gatsby-source-prismic/blob/bff33e989d25f8c5ed822efdc9b3778e447c9120/packages/gatsby-source-prismic/src/builders/buildImageThumbnailType.ts#L16

Reply:
The base/shared GraphQL types for the plugin are created here: https://github.com/angeloashmore/gatsby-source-prismic/blob/bff33e989d25f8c5ed822efdc9b3778e447c9120/packages/gatsby-source-prismic/src/create-schema-customization.ts#L30-L34

Is there a way to separate the base Imgix types from the fields? This way the types are only created once instead of anytime an image field's type is created (could be many times depending on the repository). Technically it doesn't matter since `graphql-compose` will merge everything, but just figured I'd ask.

Reply:
> Is there a way to separate the base Imgix types from the fields? This way the types are only created once instead of anytime an image field's type is created (could be many times depending on the repository).

Not with the current API design. Technically as long as the same `namespace` is passed every time, all the types will work together (thanks `graphql-compose`), so `createImgixTypes` could be called once in `src/create-schema-customization.ts`, with just the types consumed there and passed to `createTypes`, and then call it again when the fields are needed.

I'm a bit hesitant to refactor this to different functions unless there's a significant improvement to be had, and since I think it's beneficial to have a single API export. What do you think of this?

Reply:
That works for me! Creating the base types where the other base types are created in `packages/gatsby-source-prismic/src/create-schema-customization.ts` would be preferrable. The namespace would be consistent so it should work as you described.

I just commented a code change below for `packages/gatsby-source-prismic/src/create-schema-customization.ts`, but if we go this route, it could actually be reverted to the original code. Then `src/builders/buildImgixImageTypes.ts` could be edited instead to return an array of types to create.